src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java

```diff
@@ -68,28 +68,49 @@ public PublishRequestHandler(
 
     public StreamObserver<FlowUnitMessage> getClientStream(
             final StreamObserver<PublishResponse> serviceResponse) {
-        upstreamResponseStreamList.add(serviceResponse);
-        return new SendDataClientStreamUpdateConsumer(serviceResponse);
+        SendDataClientStreamUpdateConsumer streamUpdateConsumer =
+                new SendDataClientStreamUpdateConsumer(serviceResponse);
+        dataClientStreamList.add(streamUpdateConsumer);
+        return streamUpdateConsumer;
     }
 
     public void terminateUpstreamConnections() {
-        for (final StreamObserver<PublishResponse> responseStream : upstreamResponseStreamList) {
-            responseStream.onNext(
-                    PublishResponse.newBuilder()
-                            .setDataStatus(PublishResponseStatus.NODE_SHUTDOWN)
-                            .build());
-            responseStream.onCompleted();
+        for (final SendDataClientStreamUpdateConsumer streamUpdateConsumer : dataClientStreamList) {
+            StreamObserver<PublishResponse> responseStream =
+                    streamUpdateConsumer.getServiceResponse();
+            // Check whether the desired stream was already closed by the client before invoking
+            // onNext.
+            // In case stream was closed/completed by client, calling onNext will result in an
+            // exception.
+            if (!streamUpdateConsumer.isCompleted()) {
+                responseStream.onNext(
+                        PublishResponse.newBuilder()
+                                .setDataStatus(PublishResponseStatus.NODE_SHUTDOWN)
+                                .build());
+                responseStream.onCompleted();
+            }
         }
+        dataClientStreamList.clear();
+    }
+
+    // Visible for testing
+    protected List<SendDataClientStreamUpdateConsumer> getDataClientStreamList() {
+        return this.dataClientStreamList;
     }
 
-    private class SendDataClientStreamUpdateConsumer implements StreamObserver<FlowUnitMessage> {
+    protected class SendDataClientStreamUpdateConsumer implements StreamObserver<FlowUnitMessage> {
 
         private final StreamObserver<PublishResponse> serviceResponse;
+        private boolean isCompleted;
```

Comment:
Let's make the variable `final`. 

Also, would recommend renaming it to `isClosed`.

Reply:
Actually can't make it final as will then have to initialize it to a default false value. And then I won't be able to change its value later on as I am doing below.
I used isCompleted as per grpc stream terminology where we usually "complete" the response stream in the end. Though it is similar to close.

Let me know if you still think that we should rename it.

Reply:
Makes sense, we can keep the naming. 

I wanted to confirm one more thing, is the `isCompleted` variable value updated only by single owner thread ? Do we need `AtomicBoolean` here?

Reply:
Don't need a AtomicBoolean here as this object is created per node per vertex.

Reply:
LGTM then.